File: server/routes/misc.ts
import type { Hono } from 'hono'
import { db } from '../db'
import { blockPair, unblockPair } from '../matchmaking'
import { userFromToken, publicUser, type UserRow } from '../auth'
import { rateLimit } from '../rateLimit'
import { getBearer, clientIp } from '../http'
import { inc } from '../metrics'
import { noteReport } from '../alerts'
import {
  API_ROUTES,
  HTTP_STATUS,
  METRIC_NAMES,
  REPORT_REASONS,
  REPORT_STATUS_FILTER,
} from '../../shared/constants'
import type { ReportReason } from '../../shared/types'

export function registerMiscRoutes(app: Hono) {
  // Blocks
  app.post(API_ROUTES.blocks, async (c) => {
    const user = await userFromToken(getBearer(c))
    if (!user) return c.json({ error: 'Unauthorized' }, HTTP_STATUS.unauthorized)
    const { blockedId } = await c.req.json<{ blockedId?: number }>()
    if (!blockedId || blockedId === user.id) return c.json({ error: 'Invalid target' }, HTTP_STATUS.badRequest)
    await db.execute({
      sql: 'INSERT OR IGNORE INTO blocks (blocker_id, blocked_id) VALUES (?, ?)',
      args: [user.id, blockedId],
    })
    blockPair(user.id, blockedId)
    return c.json({ ok: true })
  })

  app.get(API_ROUTES.blocks, async (c) => {
    const user = await userFromToken(getBearer(c))
    if (!user) return c.json({ error: 'Unauthorized' }, HTTP_STATUS.unauthorized)
    const result = await db.execute({
      sql: `SELECT b.blocked_id AS id, u.email, b.created_at
            FROM blocks b
            LEFT JOIN users u ON u.id = b.blocked_id
            WHERE b.blocker_id = ?
            ORDER BY b.id DESC`,
      args: [user.id],
    })
    return c.json({
      blocked: result.rows.map((r) => ({
        id: Number(r.id),
        email: typeof r.email === 'string' ? r.email : null,
        createdAt: typeof r.created_at === 'string' ? r.created_at : null,
      })),
    })
  })

  app.delete(API_ROUTES.blockById(':id'), async (c) => {
    const user = await userFromToken(getBearer(c))
    if (!user) return c.json({ error: 'Unauthorized' }, HTTP_STATUS.unauthorized)
    const blockedId = Number(c.req.param('id'))
    if (!blockedId) return c.json({ error: 'Invalid id' }, HTTP_STATUS.badRequest)
    await db.execute({
      sql: 'DELETE FROM blocks WHERE blocker_id = ? AND blocked_id = ?',
      args: [user.id, blockedId],
    })
    unblockPair(user.id, blockedId)
    return c.json({ ok: true })
  })

  // Reports
  app.post(API_ROUTES.reports, async (c) => {
    const ip = clientIp(c)
    if (!rateLimit(`report:${ip}`, 15, 60_000)) return c.json({ error: 'Too many reports' }, HTTP_STATUS.tooManyRequests)
    const user = await userFromToken(getBearer(c))
    const body = await c.req.json<{ reason?: ReportReason; detail?: string; roomId?: string }>()
    const reasons: ReportReason[] = [...REPORT_REASONS]
    if (!body.reason || !reasons.includes(body.reason)) return c.json({ error: 'Invalid reason' }, HTTP_STATUS.badRequest)
    await db.execute({
      sql: 'INSERT INTO reports (reporter_id, reporter_session, room_id, reason, detail) VALUES (?, ?, ?, ?, ?)',
      args: [user?.id ?? null, ip, body.roomId ?? null, body.reason, body.detail?.slice(0, 500) ?? null],
    })
    inc(METRIC_NAMES.reportsTotal)
    void noteReport(body.reason)
    return c.json({ ok: true })
  })

  // Ratings
  app.post(API_ROUTES.ratings, async (c) => {
    const ip = clientIp(c)
    if (!rateLimit(`rating:${ip}`, 40, 60_000)) return c.json({ error: 'Too many requests' }, HTTP_STATUS.tooManyRequests)
    const user = await userFromToken(getBearer(c))
    const body = await c.req.json<{ roomId?: string; score?: number }>()
    const score = Number(body.score)
    if (!Number.isInteger(score) || score < 1 || score > 5) {
      return c.json({ error: 'Score must be 1–5.' }, HTTP_STATUS.badRequest)
    }
    const roomId = body.roomId?.slice(0, 64) || `anon_${ip}_${Date.now()}`
    try {
      await db.execute({
        sql: 'INSERT INTO ratings (room_id, rater_id, rater_session, score) VALUES (?, ?, ?, ?)',
        args: [roomId, user?.id ?? null, ip, score],
      })
    } catch {
      return c.json({ error: 'Already rated this match.' }, HTTP_STATUS.conflict)
    }
    inc(METRIC_NAMES.ratingsTotal)
    inc(`rating_score_${score}`)
    return c.json({ ok: true })
  })

  // Users search
  app.get(API_ROUTES.usersSearch, async (c) => {
    const user = await userFromToken(getBearer(c))
    if (!user) return c.json({ error: 'Unauthorized' }, HTTP_STATUS.unauthorized)
    const email = c.req.query('email')?.trim()
    if (!email) return c.json({ error: 'Email required' }, HTTP_STATUS.badRequest)
    const result = await db.execute({
      sql: 'SELECT id, email, birth_date, gender, country, language, interests, email_verified FROM users WHERE email = ? AND id != ?',
      args: [email, user.id],
    })
    const row = result.rows[0]
    if (!row) return c.json({ user: null })
    return c.json({ user: publicUser(row as unknown as UserRow) })
  })
}

File: rust/src/routes/misc.rs
//! Blocks, reports, ratings and user search. Port of `server/routes/misc.ts`.

use axum::extract::{Path, Query, State};
use axum::http::HeaderMap;
use axum::routing::{delete, get, post};
use axum::{Json, Router};
use libsql::params;
use serde::Deserialize;
use serde_json::{json, Value};

use crate::auth::session::{public_user, user_from_token, UserRow};
use crate::error::{ApiError, ApiResult};
use crate::infra::http::{client_ip, get_bearer};
use crate::infra::metrics::inc;
use crate::infra::rate_limit::rate_limit;
use crate::AppState;

const REPORT_REASONS: &[&str] = &[
    "nudity",
    "harassment",
    "hate",
    "spam",
    "underage",
    "violence",
    "other",
];

pub fn router(state: AppState) -> Router {
    Router::new()
        .route("/api/v1/blocks", post(create_block).get(list_blocks))
        .route("/api/v1/blocks/{id}", delete(remove_block))
        .route("/api/v1/reports", post(create_report))
        .route("/api/v1/ratings", post(create_rating))
        .route("/api/v1/users/search", get(search_users))
        .with_state(state)
}

async fn require_user(state: &AppState, headers: &HeaderMap) -> ApiResult<UserRow> {
    user_from_token(&state.db, get_bearer(headers).as_deref())
        .await
        .map_err(ApiError::from)?
        .ok_or_else(ApiError::unauthorized)
}

async fn create_block(
    State(state): State<AppState>,
    headers: HeaderMap,
    Json(body): Json<Value>,
) -> ApiResult<Json<Value>> {
    let user = require_user(&state, &headers).await?;
    let blocked_id = body.get("blockedId").and_then(Value::as_i64).unwrap_or(0);
    if blocked_id == 0 || blocked_id == user.id {
        return Err(ApiError::bad_request("Invalid target"));
    }
    state
        .db
        .conn()
        .execute(
            "INSERT OR IGNORE INTO blocks (blocker_id, blocked_id) VALUES (?, ?)",
            params![user.id, blocked_id],
        )
        .await?;
    // The in-memory blocked-pair set is hydrated by matchmaking in Phase 5;
    // the row above is the durable half and is already correct.
    Ok(Json(json!({ "ok": true })))
}

async fn list_blocks(State(state): State<AppState>, headers: HeaderMap) -> ApiResult<Json<Value>> {
    let user = require_user(&state, &headers).await?;
    let mut rows = state
        .db
        .conn()
        .query(
            "SELECT b.blocked_id AS id, u.email, b.created_at
             FROM blocks b
             LEFT JOIN users u ON u.id = b.blocked_id
             WHERE b.blocker_id = ?
             ORDER BY b.id DESC",
            params![user.id],
        )
        .await?;
    let mut blocked = Vec::new();
    while let Some(row) = rows.next().await? {
        blocked.push(json!({
            "id": row.get::<i64>(0).unwrap_or(0),
            "email": row.get::<String>(1).ok(),
            "createdAt": row.get::<String>(2).ok(),
        }));
    }
    Ok(Json(json!({ "blocked": blocked })))
}

async fn remove_block(
    State(state): State<AppState>,
    headers: HeaderMap,
    Path(id): Path<i64>,
) -> ApiResult<Json<Value>> {
    let user = require_user(&state, &headers).await?;
    if id == 0 {
        return Err(ApiError::bad_request("Invalid id"));
    }
    state
        .db
        .conn()
        .execute(
            "DELETE FROM blocks WHERE blocker_id = ? AND blocked_id = ?",
            params![user.id, id],
        )
        .await?;
    Ok(Json(json!({ "ok": true })))
}

async fn create_report(
    State(state): State<AppState>,
    headers: HeaderMap,
    Json(body): Json<Value>,
) -> ApiResult<Json<Value>> {
    let ip = client_ip(&headers);
    if !rate_limit(&format!("report:{ip}"), 15, 60_000) {
        return Err(ApiError::too_many("Too many reports"));
    }
    // Reports are accepted from guests too, so the user lookup is optional.
    let user = user_from_token(&state.db, get_bearer(&headers).as_deref())
        .await
        .map_err(ApiError::from)?;

    let reason = body.get("reason").and_then(Value::as_str).unwrap_or_default();
    if !REPORT_REASONS.contains(&reason) {
        return Err(ApiError::bad_request("Invalid reason"));
    }
    // `detail?.slice(0, 500)` — slice counts UTF-16 units in JS; chars here.
    let detail: Option<String> = body
        .get("detail")
        .and_then(Value::as_str)
        .map(|d| d.chars().take(500).collect());

    state
        .db
        .conn()
        .execute(
            "INSERT INTO reports (reporter_id, reporter_session, room_id, reason, detail) VALUES (?, ?, ?, ?, ?)",
            params![
                user.as_ref().map(|u| u.id),
                ip,
                body.get("roomId").and_then(Value::as_str),
                reason,
                detail
            ],
        )
        .await?;
    inc("reports_total", 1);
    crate::alerts::note_report(reason).await;
    Ok(Json(json!({ "ok": true })))
}

async fn create_rating(
    State(state): State<AppState>,
    headers: HeaderMap,
    Json(body): Json<Value>,
) -> ApiResult<Json<Value>> {
    let ip = client_ip(&headers);
    if !rate_limit(&format!("rating:{ip}"), 40, 60_000) {
        return Err(ApiError::too_many("Too many requests"));
    }
    let user = user_from_token(&state.db, get_bearer(&headers).as_deref())
        .await
        .map_err(ApiError::from)?;

    // `Number.isInteger(score)` — a fractional score is rejected, not rounded.
    let raw = body.get("score").and_then(Value::as_f64).unwrap_or(f64::NAN);
    if !raw.is_finite() || raw.fract() != 0.0 || raw < 1.0 || raw > 5.0 {
        return Err(ApiError::bad_request("Score must be 1–5."));
    }
    let score = raw as i64;

    let room_id = body
        .get("roomId")
        .and_then(Value::as_str)
        .map(|s| s.chars().take(64).collect::<String>())
        .filter(|s| !s.is_empty())
        .unwrap_or_else(|| format!("anon_{ip}_{}", now_ms()));

    // UNIQUE(room_id, rater_session) turns a second rating into a conflict.
    if state
        .db
        .conn()
        .execute(
            "INSERT INTO ratings (room_id, rater_id, rater_session, score) VALUES (?, ?, ?, ?)",
            params![room_id, user.as_ref().map(|u| u.id), ip, score],
        )
        .await
        .is_err()
    {
        return Err(ApiError::conflict("Already rated this match."));
    }

    inc("ratings_total", 1);
    inc(&format!("rating_score_{score}"), 1);
    Ok(Json(json!({ "ok": true })))
}

fn now_ms() -> u64 {
    std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map(|d| d.as_millis() as u64)
        .unwrap_or(0)
}

#[derive(Deserialize)]
struct EmailQuery {
    email: Option<String>,
}

async fn search_users(
    State(state): State<AppState>,
    headers: HeaderMap,
    Query(q): Query<EmailQuery>,
) -> ApiResult<Json<Value>> {
    let user = require_user(&state, &headers).await?;
    let email = q
        .email
        .as_deref()
        .map(str::trim)
        .filter(|s| !s.is_empty())
        .ok_or_else(|| ApiError::bad_request("Email required"))?;

    let mut rows = state
        .db
        .conn()
        .query(
            "SELECT id, email, birth_date, gender, country, language, interests, email_verified
             FROM users WHERE email = ? AND id != ?",
            params![email, user.id],
        )
        .await?;
    let Some(row) = rows.next().await? else {
        return Ok(Json(json!({ "user": Value::Null })));
    };
    let found = UserRow {
        id: row.get(0)?,
        email: row.get(1)?,
        birth_date: row.get(2).ok(),
        gender: row.get(3).ok(),
        country: row.get(4).ok(),
        language: row.get(5).ok(),
        interests: row.get(6).ok(),
        email_verified: row.get(7).unwrap_or(0),
    };
    Ok(Json(json!({ "user": public_user(&found) })))
}
